File: app/src/main/java/net/gmx/szermatt/dreamcatcher/PowerOffWorker.java
package net.gmx.szermatt.dreamcatcher;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import net.gmx.szermatt.dreamcatcher.harmony.PowerOffTask;

/** Wraps a {@link PowerOffTask} into a Worker. */
public class PowerOffWorker extends Worker {
    private static final String TAG = "DreamCatcher";

    private final PowerOffTask task = new PowerOffTask();

    public PowerOffWorker(
            @NonNull Context context,
            @NonNull WorkerParameters params) {
        super(context, params);
    }

    @Override
    public Result doWork() {
        try {
            Log.i(TAG, "PowerOffWorker launched");
            task.run();
            Log.i(TAG, "PowerOffWorker succeeded");
            return Result.success();
        } catch (Exception e) {
            Log.e(TAG, "PowerOffWorker failed", e);
            return Result.failure();
        }
    }

    @Override
    public void onStopped() {
        Log.w(TAG, "PowerOffWorker stopped");
        task.stop();
    }
}

File: app/src/main/kotlin/net/gmx/szermatt/dreamcatcher/PowerOffWorker.kt
package net.gmx.szermatt.dreamcatcher

import android.content.Context
import android.util.Log
import androidx.annotation.GuardedBy
import androidx.work.*
import kotlinx.coroutines.ExperimentalCoroutinesApi
import net.gmx.szermatt.dreamcatcher.DreamcatcherApplication.Companion.TAG
import net.gmx.szermatt.dreamcatcher.harmony.PowerOffTask
import java.util.concurrent.CancellationException
import java.util.concurrent.TimeUnit

/** Wraps a [PowerOffTask] into a Worker.  */
class PowerOffWorker(
    context: Context,
    params: WorkerParameters
) : CoroutineWorker(context, params) {

    companion object {
        /** Creates a one-time [WorkRequest] for this worker. */
        fun workRequest(
            uuid: String? = null,
            delayInMinutes: Int = 0,
            dryRun: Boolean = false,
            tag: String? = null
        ): WorkRequest {
            val b = OneTimeWorkRequest.Builder(PowerOffWorker::class.java)
            if (delayInMinutes > 0) {
                b.setConstraints(
                    Constraints.Builder()
                        .setRequiredNetworkType(NetworkType.CONNECTED)
                        .build()
                )
                b.setInitialDelay(delayInMinutes.toLong(), TimeUnit.MINUTES)
            }
            tag?.let { b.addTag(it) }
            val data = Data.Builder()
            data.putBoolean("dryRun", dryRun)
            data.putString("uuid", uuid)
            b.setInputData(data.build())
            return b.build()
        }
    }

    @GuardedBy("this")
    private val task = PowerOffTask(
        listener = object : PowerOffTask.Listener {
            override fun onPowerOffTaskProgress(step: Int, stepCount: Int) {
                val data = Data.Builder()
                WorkProgressFragment.fillProgressData(data, step, stepCount)
                setProgressAsync(data.build())
            }
        }
    )

    @ExperimentalCoroutinesApi
    override suspend fun doWork(): Result {
        return try {
            val dryRun = inputData.getBoolean("dryRun", false)
            Log.d(TAG, "PowerOffWorker launched dryRun=$dryRun ")
            task.run(uuid = inputData.getString("uuid"), dryRun = dryRun)
            Log.d(TAG, "PowerOffWorker succeeded")
            Result.success()
        } catch (e: CancellationException) {
            Log.d(TAG, "PowerOffWorker finished after being stopped", e)
            Result.success()
        } catch (e: Exception) {
            Log.e(TAG, "PowerOffWorker failed with exception ${e}", e)
            Result.failure()
        }
    }
}